Libraries/LibTLS/ClientHandshake.cpp

```diff
@@ -0,0 +1,614 @@
+        return (i8)Error::NeedMoreData;
+    }
+
+    size_t index = 3;
+
+    u32 size = buffer[0] * 0x10000 + buffer[1] * 0x100 + buffer[2];
+    index += 3;
+
+    if (size < 12) {
+        dbg() << "finished packet smaller than minimum size: " << size;
+        return (i8)Error::BrokenPacket;
+    }
+
+    if (size < buffer.size() - index) {
+        dbg() << "not enough data after length: " << size << " > " << buffer.size() - index;
+        return (i8)Error::NeedMoreData;
+    }
+
+    // TODO: Compare Hashes
+    dbg() << "FIXME: handle_finished :: Check message validity";
+    m_context.connection_status = 0xff;
+
+    return handle_message(buffer);
+}
+
+void TLSv12::build_random(PacketBuilder& builder)
+{
+    u8 random_bytes[48];
+    size_t bytes = 48;
+
+    arc4random_buf(random_bytes, bytes);
+
+    // remove zeros from the random bytes
+    for (size_t i = 0; i < bytes; ++i)
+        if (!random_bytes[i])
+            random_bytes[i--] = arc4random();
+
+    if (m_context.is_server) {
+        dbg() << "Server mode not supported";
+        return;
+    } else {
+        *(u16*)random_bytes = convert_between_host_and_network((u16)Version::V12);
+    }
+
+    m_context.premaster_key = ByteBuffer::copy(random_bytes, bytes);
+
+    const auto& certificate = m_context.certificates[0];
+#ifdef TLS_DEBUG
+    dbg() << "PreMaster secret";
+    print_buffer(m_context.premaster_key);
+#endif
+
+    Crypto::PK::RSA_PKCS1_EME rsa(certificate.public_key.modulus(), 0, certificate.public_key.public_exponent());
+
+    u8 out[rsa.output_size()];
+    auto outbuf = ByteBuffer::wrap(out, rsa.output_size());
+    rsa.encrypt(m_context.premaster_key, outbuf);
+
+#ifdef TLS_DEBUG
+    dbg() << "Encrypted: ";
+    print_buffer(outbuf);
+#endif
+
+    if (!compute_master_secret(bytes)) {
+        dbg() << "oh noes we could not derive a master key :(";
+        return;
+    }
+
+    builder.append_u24(outbuf.size() + 2);
+    builder.append((u16)outbuf.size());
+    builder.append(outbuf);
+}
+
+ssize_t TLSv12::handle_payload(const ByteBuffer& vbuffer)
+{
+    if (m_context.connection_status == 0xff) {
+        auto packet = build_alert(false, (u8)AlertDescription::NoRenegotiation);
+        write_packet(packet);
+        return 1;
+    }
+    auto buffer = vbuffer;
+    auto buffer_length = buffer.size();
+    auto original_length = buffer_length;
+    while (buffer_length >= 4 && !m_context.critical_error) {
+        ssize_t payload_res = 0;
+        if (buffer_length < 1)
+            return (i8)Error::NeedMoreData;
+        auto type = buffer[0];
+        size_t write_packets = 0;
+        size_t payload_size = buffer[1] * 0x10000 + buffer[2] * 0x100 + buffer[3] + 3;
+#ifdef TLS_DEBUG
+        dbg() << "payload size: " << payload_size << " buffer length: " << buffer_length;
+#endif
+        if (payload_size + 1 > buffer_length)
+            return (i8)Error::NeedMoreData;
+
+        switch (type) {
+        // hello request
+        case 0x00:
```

Comment:
Styling, feel free to ignore : Rather than having comments + magic values in the code, maybe add some `#DEFINE` or constexprs ?